lib/Parse/Parser.cpp

```diff
@@ -1016,7 +1017,7 @@ Parser::parseList(tok RightK, SourceLoc LeftLoc, SourceLoc &RightLoc,
     }
     // If we're in a comma-separated list, the next token is at the
     // beginning of a new line and can never start an element, break.
-    if (Tok.isAtStartOfLine() &&
+    if (Tok.isAtStartOfLine() && !AllowSepOmission &&
```

Comment:
I'm not sure.  My goal with this change was to enable commas to be omitted from argument lists whose argument labels are contextual keywords.  Is there a better way to achieve that?

Reply:
What do you mean by contextual keywords? I'm just trying to point out if there is an element on the new line but it's something which is not allowed to be in the list in general e.g. declaration or statement, loop should stop just like before. Comma omitting should only be allowed if next token represents something which could syntactically belong in the list...

Reply:
`protocol`, for example, can be used as an argument label:

```swift
func foo(a: Int, protocol: Int, c: Int) {}
```

It would be nice to be able to omit the comma before the second item in the list `protocol: 16`:

```swift
foo(
    a: 13
    protocol: 16
    c: 15
)

```

Without `!AllowSepOmission &&` or a better change, that fails to compile because `isStartOfDecl()` returns true.

```
main.swift:28:5: error: expected ')' in expression list
    protocol: 16
    ^
main.swift:26:4: note: to match this opening '('
foo(
   ^
main.swift:27:10: error: missing argument for parameter 'protocol' in call
    a: 13
         ^
         , protocol: <#Int#>
main.swift:24:6: note: 'foo(a:protocol:)' declared here
func foo(a: Int, protocol: Int) {}
     ^
main.swift:28:13: error: expected identifier in protocol declaration
    protocol: 16
            ^
```

Another approach would be to modify `isStartOfDecl()` to look ahead farther.  Would that be a better solution?  It seems like it might be enough to peek one token after the keyword and if it's a colon, return `false`--looking at the list of keywords in `isKeywordPossibleDeclStart()`, it seems that none of them can be followed immediately by a colon.  Is that right?

That alternate change corrects the regression in `test/decl/subscript/subscripting.swift`.

Reply:
I definitely see your point that if the item in the list is 

> something which is not allowed to be in the list in general e.g. declaration or statement, loop should stop just like before

so it seems that enhancing `isStartOfDecl()` and `isStartOfStmt()` to be slightly more discerning seems like a much better solution.  The loops stops like before when the next item is a declaration or statement.

Reply:
Fixed in https://github.com/apple/swift/pull/22714/commits/acdd16f7d16d1d2223758b09f5327175f1c3f8d6 .

Reply:
I agree that the best solution, as you suggested, would be indeed to do  lookahead at what is going on to identify label syntax. It seems like `isStartOfDecl` should know that `identifier:` doesn't constitute a valid declaration (unless it starts with `case`). There is  `canBeArgumentLabel()` method which could be used in combination with `peekToken.is(tok::colon)` just like what `Parser::parseOptionalArgumentLabel` does to determine that.

Reply:
Ok, as of https://github.com/apple/swift/pull/22714/commits/faecb8d5f15a3421f661bd606158d0394dbb19be , `canBeArgumentLabel()` is used alongside `peekToken().is(tok::colon)`, and it's all done in `parseList`.